fix(opentelemetry_cowboy): prevent telemetry handler crash on invalid HTTP/2 scheme

velimir opened the pull request:
Cowboy does not validate the `:scheme` pseudo-header per RFC 9113, which
states that scheme "is not restricted to http and https" and can be used
for non-HTTP schemes via proxies/gateways. This means Cowboy passes through
any value from the HTTP/2 HEADERS frame without validation.

When a malformed or unexpected scheme value reaches `extract_scheme/2`,
the case clause crashes because it only handles `<<"http">>` and
`<<"https">>`. This crash causes Erlang's telemetry library to permanently
detach the handler for the affected node, resulting in all subsequent
requests losing OpenTelemetry tracing until the node is restarted.

In production, this manifests as "almost all" requests suddenly missing
trace IDs, with the root cause being a single malformed HTTP/2 request
that crashed the telemetry handler.

The fix introduces a configurable `valid_schemes` option that maps scheme
binaries to atoms. Unknown schemes now return `undefined` instead of
crashing, which is consistent with how `otel_http:extract_scheme/2` already
handles unknown schemes from headers.

The default configuration preserves backwards compatibility by only
accepting `http` and `https`. Users who need to support additional schemes
(e.g., `ws`, `wss`) can extend the map via configuration.

Fixes: #567

velimir committed:
upgrade opentelemetry_api to 1.5.0

This to fix the following compilation error on test:

```
$ rebar3 ct
===> Verifying dependencies...
===> Analyzing applications...
===> Compiling opentelemetry
===> Compiling _build/test/lib/opentelemetry/src/otel_span_utils.erl failed
    ┌─ _build/test/lib/opentelemetry/src/otel_span_utils.erl:
    │
 92 │              {ParentSpanCtx#span_ctx{span_id=SpanId, hex_span_id=HexSpanId}, ParentSpanId, ParentIsRemote}
    │                                                      ╰── field hex_span_id undefined in record span_ctx

     ┌─ _build/test/lib/opentelemetry/src/otel_span_utils.erl:
     │
 101 │                hex_trace_id=HexTraceId,
     │                ╰── field hex_trace_id undefined in record span_ctx

     ┌─ _build/test/lib/opentelemetry/src/otel_span_utils.erl:
     │
 103 │                hex_span_id=HexSpanId,
     │                ╰── field hex_span_id undefined in record span_ctx

    ┌─ _build/test/lib/opentelemetry/src/otel_span_utils.erl:
    │
 91 │              HexSpanId = otel_utils:encode_hex(<<SpanId:64>>),
    │              ╰── Warning: variable 'HexSpanId' is unused

    ┌─ _build/test/lib/opentelemetry/src/otel_span_utils.erl:
    │
 97 │      HexTraceId = otel_utils:encode_hex(<<TraceId:128>>),
    │      ╰── Warning: variable 'HexTraceId' is unused

    ┌─ _build/test/lib/opentelemetry/src/otel_span_utils.erl:
    │
 99 │      HexSpanId = otel_utils:encode_hex(<<SpanId:64>>),
    │      ╰── Warning: variable 'HexSpanId' is unused
```

this is necessary as there are new fields got introduced in the the https://github.com/open-telemetry/opentelemetry-erlang/pull/874

velimir committed:
fix(opentelemetry_cowboy): prevent telemetry handler crash on invalid HTTP/2 scheme

Cowboy does not validate the `:scheme` pseudo-header per RFC 9113, which
states that scheme "is not restricted to http and https" and can be used
for non-HTTP schemes via proxies/gateways. This means Cowboy passes through
any value from the HTTP/2 HEADERS frame without validation.

When a malformed or unexpected scheme value reaches `extract_scheme/2`,
the case clause crashes because it only handles `<<"http">>` and
`<<"https">>`. This crash causes Erlang's telemetry library to permanently
detach the handler for the affected node, resulting in all subsequent
requests losing OpenTelemetry tracing until the node is restarted.

In production, this manifests as "almost all" requests suddenly missing
trace IDs, with the root cause being a single malformed HTTP/2 request
that crashed the telemetry handler.

The fix introduces a configurable `valid_schemes` option that maps scheme
binaries to atoms. Unknown schemes now return `undefined` instead of
crashing, which is consistent with how `otel_http:extract_scheme/2` already
handles unknown schemes from headers.

The default configuration preserves backwards compatibility by only
accepting `http` and `https`. Users who need to support additional schemes
(e.g., `ws`, `wss`) can extend the map via configuration.

Reproduction repository: https://github.com/velimir/scheme_crash

Fixes: https://github.com/open-telemetry/opentelemetry-erlang-contrib/issues/567

velimir commented on instrumentation/opentelemetry_cowboy/test/opentelemetry_cowboy_SUITE.erl:
```diff
@@ -507,3 +508,20 @@ binary_status_code_request(_Config) ->
+            ?assertEqual(undefined, maps:get(?URL_SCHEME, otel_attributes:map(Attributes)))
+    after
+        1000 -> ct:fail(invalid_scheme_returns_undefined)
+    end.
```
I would love to, but making an http2 request via `httpc` is.. problematic.

velimir commented:
hey @tsloughter is there anything I can help you with to get this merged?

tsloughter commented on instrumentation/opentelemetry_cowboy/src/opentelemetry_cowboy.erl:
```diff
@@ -95,7 +95,8 @@ default_opts() ->
         server_address_headers => [<<"forwarded">>, <<"x-forwarded-host">>, <<"host">>],
-        server_address_headers_sort_fn => undefined
+        server_address_headers_sort_fn => undefined,
+        valid_schemes => #{<<"http">> => http, <<"https">> => https}
```
Oh, invalid schemas? How were those getting through... interesting

velimir committed:
Merge remote-tracking branch 'origin/main' into fix-scheme-crash

velimir committed:
fix(opentelemetry_cowboy): use raw scheme binary instead of atom conversion

Remove valid_schemes config and pass Cowboy's ReqScheme binary through
directly, avoiding the need for validation/conversion that could crash
on unexpected scheme values from malformed HTTP/2 requests.

velimir commented on instrumentation/opentelemetry_cowboy/src/opentelemetry_cowboy.erl:
```diff
@@ -236,12 +239,7 @@ extract_scheme(Req, Config) ->
-                <<"https">> ->
-                    https
-            end;
+            maps:get(ReqScheme, ValidSchemes, undefined);
```
Done in 86e592c — removed `valid_schemes` entirely and pass `ReqScheme` binary through directly.

velimir commented:
@tsloughter this is ready for another look when you get a chance

tsloughter commented:
Hey, sorry, I should have reviewed closer the first time. We don't just want to return `ReqScheme` since that could be used to blow up the cardinality of the scheme attribute. The right fix, I think, is to update `otel_http` to support `ws` and `wss`.

yordis committed:
Merge remote-tracking branch 'upstream/main' into fix-scheme-crash

yordis committed:
fix(opentelemetry_cowboy): bound scheme cardinality and pin test cowboy version

Returning the raw HTTP/2 :scheme pseudo-header unchecked would let a
client-controlled string flow into url.scheme with unbounded cardinality,
since Cowboy does not validate it per RFC 9113.

The test profile's cowboy constraint floated to any 2.x release, so newer
cowboy versions requiring a newer cowlib than the locked default profile
broke CI dependency resolution under rebar3 3.24.

Signed-off-by: Yordis Prieto <yordis.prieto@gmail.com>
